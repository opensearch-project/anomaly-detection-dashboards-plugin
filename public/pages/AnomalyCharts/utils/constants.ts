/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

export const INITIAL_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0.0,
  maxAnomalyGrade: 0.0,
  minConfidence: 0.0,
  maxConfidence: 0.0,
  lastAnomalyOccurrence: '',
};

export enum CHART_FIELDS {
  PLOT_TIME = 'plotTime',
  ANOMALY_GRADE = 'anomalyGrade',
  CONFIDENCE = 'confidence',
  DATA = 'data',
  AGG_INTERVAL = 'aggInterval',
  EXPECTED_VALUE = 'expectedValue',
}

export enum CHART_COLORS {
  ANOMALY_GRADE_COLOR = '#D13212',
  FEATURE_DATA_COLOR = '#16191F',
  FEATURE_COLOR = '#fcd529',
  CONFIDENCE_COLOR = '#017F75',
  LIGHT_BACKGROUND = '#FFFFFF',
  DARK_BACKGROUND = '#1D1E24',
}

export const FEATURE_CHART_THEME = [
  {
    lineSeriesStyle: {
      line: {
        strokeWidth: 2,
        visible: true,
        opacity: 0.5,
      },
      point: {
        visible: true,
        stroke: CHART_COLORS.FEATURE_DATA_COLOR,
      },
    },
  },
  {
    colors: {
      vizColors: [CHART_COLORS.FEATURE_DATA_COLOR],
    },
  },
];

export const ANOMALY_CHART_THEME = [
  {
    colors: {
      vizColors: [
        CHART_COLORS.CONFIDENCE_COLOR,
        CHART_COLORS.ANOMALY_GRADE_COLOR,
      ],
    },
  },
];

export const DATE_PICKER_QUICK_OPTIONS = [
  { start: 'now-24h', end: 'now', label: 'last 24 hours' },
  { start: 'now-7d', end: 'now', label: 'last 7 days' },
  { start: 'now-30d', end: 'now', label: 'last 30 days' },
  { start: 'now-90d', end: 'now', label: 'last 90 days' },

  { start: 'now/d', end: 'now', label: 'Today' },
  { start: 'now/w', end: 'now', label: 'Week to date' },
  { start: 'now/M', end: 'now', label: 'Month to date' },
  { start: 'now/y', end: 'now', label: 'Year to date' },
];

export enum LIVE_CHART_CONFIG {
  REFRESH_INTERVAL_IN_SECONDS = 30 * 1000, //poll anomaly result every 30 seconds
  MONITORING_INTERVALS = 60,
}

export const DEFAULT_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0,
  maxAnomalyGrade: 0,
  minConfidence: 0,
  maxConfidence: 0,
  lastAnomalyOccurrence: '-',
};

export const HEATMAP_CHART_Y_AXIS_WIDTH = 30;
