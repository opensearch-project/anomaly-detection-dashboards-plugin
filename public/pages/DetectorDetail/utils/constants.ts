/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

export enum DETECTOR_DETAIL_TABS {
  RESULTS = 'results',
  HISTORICAL = 'historical',
  CONFIGURATIONS = 'configurations',
}

const DEFAULT_ACTION_ITEM = 'Restart the detector and try again.';
// Known causes:
// https://github.com/opensearch-project/anomaly-detection/blob/development/src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java#L174-L185
export const DETECTOR_INIT_FAILURES = Object.freeze({
  NO_TRAINING_DATA: {
    //https://github.com/opensearch-project/anomaly-detection/blob/development/src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java#L801
    keyword: 'Cannot get training data',
    cause: 'no sufficient data is ingested',
    actionItem:
      'Make sure your data is ingested correctly. If your data source has infrequent ingestion, increase the detector time interval and try again.',
  },
  COLD_START_ERROR: {
    //https://github.com/opensearch-project/anomaly-detection/blob/development/src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java#L811
    keyword: 'Error while cold start',
    cause: 'of an error during model training',
    actionItem: DEFAULT_ACTION_ITEM,
  },
  AD_MODEL_MEMORY_REACH_LIMIT: {
    //https://github.com/opensearch-project/anomaly-detection/blob/development/src/main/java/com/amazon/opendistroforelasticsearch/ad/ml/ModelManager.java#L272
    keyword: 'AD models memory usage exceeds our limit',
    cause: 'of lack of memory for the detector models',
    actionItem: 'Reduce the number of features and try again.',
  },
  DETECTOR_MEMORY_REACH_LIMIT: {
    //https://github.com/opensearch-project/anomaly-detection/blob/development/src/main/java/com/amazon/opendistroforelasticsearch/ad/ml/ModelManager.java#L783
    keyword: 'Exceeded memory limit',
    cause: 'of lack of memory for the detector',
    actionItem:
      "Remove or stop other detectors that you don't actively use, increase your cluster size, reduce the number of features, or scale up with an instance type of more memory and try again.",
  },
  DATA_INDEX_NOT_FOUND: {
    //https://github.com/opensearch-project/anomaly-detection/blob/development/src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java#L366
    keyword: 'Having trouble querying data: ',
    cause: 'the data index is not found',
    actionItem: 'Make sure your index exists and try again.',
  },
  ALL_FEATURES_DISABLED: {
    //https://github.com/opensearch-project/anomaly-detection/blob/development/src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java#L368
    keyword:
      'Having trouble querying data because all of your features have been disabled',
    cause: 'all detector features are disabled',
    actionItem: 'Enable one or more features and try again.',
  },
  DETECTOR_UNDEFINED: {
    //https://github.com/opensearch-project/anomaly-detection/blob/development/src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java#L230
    keyword: 'AnomalyDetector is not available',
    cause: 'the detector is not defined',
    actionItem: 'Define your detector and try again.',
  },
  INVALID_FEATURE_QUERY: {
    // TODO: https://github.com/opensearch-project/anomaly-detection/issues/218
    //https://github.com/opensearch-project/anomaly-detection/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/ad/constant/CommonErrorMessages.java#L28
    keyword: 'Invalid search query',
    cause: 'the custom query caused an invalid feature aggregation',
    actionItem:
      'Revise the custom query in your feature aggregations and try again.',
  },
  UNKNOWN_EXCEPTION: {
    //https://github.com/opensearch-project/anomaly-detection/blob/development/src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java#L438
    keyword: 'We might have bugs',
    cause: 'of unknown error',
    actionItem: DEFAULT_ACTION_ITEM,
  },
});
