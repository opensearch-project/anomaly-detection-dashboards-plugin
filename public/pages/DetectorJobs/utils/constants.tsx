/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { DetectorJobsFormikValues } from '../../DetectorJobs/models/interfaces';

export const INITIAL_DETECTOR_JOB_VALUES: DetectorJobsFormikValues = {
  realTime: true,
  historical: false,
  startTime: 'now-30d',
  endTime: 'now',
};

export const HISTORICAL_DATE_RANGE_COMMON_OPTIONS = [
  { start: 'now-24h', end: 'now', label: 'last 24 hours' },
  { start: 'now-7d', end: 'now', label: 'last 7 days' },
  { start: 'now-30d', end: 'now', label: 'last 30 days' },
  { start: 'now-90d', end: 'now', label: 'last 90 days' },

  { start: 'now/d', end: 'now', label: 'Today' },
  { start: 'now/w', end: 'now', label: 'Week to date' },
  { start: 'now/M', end: 'now', label: 'Month to date' },
  { start: 'now/y', end: 'now', label: 'Year to date' },
];
