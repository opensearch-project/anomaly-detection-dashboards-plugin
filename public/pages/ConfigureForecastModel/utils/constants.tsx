/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  ModelConfigurationFormikValues,
} from '../../ConfigureForecastModel/models/interfaces';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

// FIXME: We intentionally leave some parameters undefined to encourage customers to click 
// the "Suggest parameters" button rather than providing default values like "10 minutes" for interval
export const INITIAL_MODEL_CONFIGURATION_VALUES: ModelConfigurationFormikValues =
  {
    shingleSize: DEFAULT_SHINGLE_SIZE,
    imputationOption: undefined,
    interval: undefined,
    windowDelay: undefined,
    suggestedSeasonality: undefined,
    recencyEmphasis: undefined,
    resultIndex: undefined,
    resultIndexMinAge: 7,
    resultIndexMinSize: 51200,
    resultIndexTtl: 60,
    flattenCustomResultIndex: false,
  };

// an enum for the sparse data handling options
export enum SparseDataOptionValue {
  IGNORE = 'ignore',
  PREVIOUS_VALUE = 'previous_value',
  SET_TO_ZERO = 'set_to_zero',
  CUSTOM_VALUE = 'custom_value',
}
