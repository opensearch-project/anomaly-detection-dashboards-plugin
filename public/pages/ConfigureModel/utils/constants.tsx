/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import {
  ModelConfigurationFormikValues,
  FeaturesFormikValues,
} from '../../ConfigureModel/models/interfaces';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

export const INITIAL_MODEL_CONFIGURATION_VALUES: ModelConfigurationFormikValues =
  {
    featureList: [],
    categoryFieldEnabled: false,
    categoryField: [],
    shingleSize: DEFAULT_SHINGLE_SIZE,
    imputationOption: undefined
  };

export const INITIAL_FEATURE_VALUES: FeaturesFormikValues = {
  featureId: '',
  featureName: '',
  featureEnabled: true,
  featureType: FEATURE_TYPE.SIMPLE,
  aggregationQuery: JSON.stringify(
    { aggregation_name: { sum: { field: 'field_name' } } },
    null,
    4
  ),
  aggregationBy: '',
  aggregationOf: [],
};

export const FEATURE_TYPES = [
  { text: 'Custom Aggregation', value: FEATURE_TYPE.CUSTOM },
  { text: 'Defined Aggregation', value: FEATURE_TYPE.SIMPLE },
];

export const FEATURE_TYPE_OPTIONS = [
  { text: 'Field value', value: FEATURE_TYPE.SIMPLE },
  { text: 'Custom expression', value: FEATURE_TYPE.CUSTOM },
];

export enum SAVE_FEATURE_OPTIONS {
  START_AD_JOB = 'start_ad_job',
  KEEP_AD_JOB_STOPPED = 'keep_ad_job_stopped',
}

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'value_count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];

export const FEATURE_FIELDS = [
  'featureName',
  'aggregationOf',
  'aggregationBy',
  'aggregationQuery',
];

// an enum for the sparse data handling options
export enum SparseDataOptionValue {
  IGNORE = 'ignore',
  PREVIOUS_VALUE = 'previous_value',
  SET_TO_ZERO = 'set_to_zero',
  CUSTOM_VALUE = 'custom_value',
}
