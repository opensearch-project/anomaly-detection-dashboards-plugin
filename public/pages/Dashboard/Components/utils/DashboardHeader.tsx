/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import {
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
  EuiPageHeader,
  EuiTitle,
} from '@elastic/eui';
import { PLUGIN_NAME, APP_PATH } from '../../../../utils/constants';
export interface DashboardHeaderProps {
  hasDetectors: boolean;
}

export const DashboardHeader = (props: DashboardHeaderProps) => {
  return (
    <EuiPageHeader>
      <EuiFlexGroup justifyContent="spaceBetween">
        <EuiFlexItem grow={false}>
          <EuiTitle size="l">
            <h1>Real-time dashboard</h1>
          </EuiTitle>
        </EuiFlexItem>
        {props.hasDetectors ? (
          <EuiFlexItem grow={false}>
            <EuiButton
              fill
              href={`${PLUGIN_NAME}#${APP_PATH.CREATE_DETECTOR}`}
              data-test-subj="add_detector"
            >
              Create detector
            </EuiButton>
          </EuiFlexItem>
        ) : null}
      </EuiFlexGroup>
    </EuiPageHeader>
  );
};
