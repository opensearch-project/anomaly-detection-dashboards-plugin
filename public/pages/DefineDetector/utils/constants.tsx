/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FILTER_TYPES, UIFilter } from '../../../models/interfaces';
import { OPERATORS_MAP } from '../../DefineDetector/components/DataFilterList/utils/constant';
import { DetectorDefinitionFormikValues } from '../../DefineDetector/models/interfaces';
export const EMPTY_UI_FILTER: UIFilter = {
  filterType: FILTER_TYPES.SIMPLE,
  fieldInfo: [],
  operator: OPERATORS_MAP.IS,
  fieldValue: '',
  query: '',
  label: '',
};

export const testDetectorDefinitionValues: DetectorDefinitionFormikValues = {
  name: 'test',
  description: 'desc',
  index: [],
  filters: [],
  filterQuery: JSON.stringify({ bool: { filter: [] } }, null, 4),
  timeField: 'timestamp',
  interval: 10,
  windowDelay: 1,
  resultIndex: undefined,
};

export const INITIAL_DETECTOR_DEFINITION_VALUES: DetectorDefinitionFormikValues =
  {
    name: '',
    description: '',
    index: [],
    filters: [],
    filterQuery: JSON.stringify({ bool: { filter: [] } }, null, 4),
    timeField: '',
    interval: 10,
    windowDelay: 1,
    resultIndex: undefined,
    resultIndexMinAge: 7,
    resultIndexMinSize: 51200,
    resultIndexTtl: 60,
    flattenCustomResultIndex: false,
    imputationMethod: undefined,
    customImputationValue: undefined
  };
