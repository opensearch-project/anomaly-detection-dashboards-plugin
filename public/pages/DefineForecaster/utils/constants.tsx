/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FILTER_TYPES, UIFilter } from '../../../models/interfaces';
import { OPERATORS_MAP } from '../../DefineDetector/components/DataFilterList/utils/constant';
import { ForecasterDefinitionFormikValues } from '../models/interfaces';
import { FEATURE_TYPE } from '../../../models/interfaces';
import {
  FeaturesFormikValues,
} from '../../ConfigureModel/models/interfaces';

export const EMPTY_UI_FILTER: UIFilter = {
  filterType: FILTER_TYPES.SIMPLE,
  fieldInfo: [],
  operator: OPERATORS_MAP.IS,
  fieldValue: '',
  query: '',
  label: '',
};

export const testForecasterDefinitionValues: ForecasterDefinitionFormikValues = {
  name: 'test',
  description: 'desc',
  index: [],
  filters: [],
  filterQuery: JSON.stringify({ bool: { filter: [] } }, null, 4),
  timeField: 'timestamp',
  featureList: [],
  categoryFieldEnabled: false,
  categoryField: [],
};

export const INITIAL_FORECASTER_DEFINITION_VALUES: ForecasterDefinitionFormikValues =
  {
    name: '',
    description: '',
    index: [],
    filters: [],
    filterQuery: JSON.stringify({ bool: { filter: [] } }, null, 4),
    timeField: '',
    featureList: [],
    categoryFieldEnabled: false,
    categoryField: [],
  };

  export const FEATURE_TYPES = [
    { text: 'Custom Aggregation', value: FEATURE_TYPE.CUSTOM },
    { text: 'Defined Aggregation', value: FEATURE_TYPE.SIMPLE },
  ];

  export const FEATURE_TYPE_OPTIONS = [
    { text: 'Field value', value: FEATURE_TYPE.SIMPLE },
    { text: 'Custom expression', value: FEATURE_TYPE.CUSTOM },
  ];

  export const INITIAL_FEATURE_VALUES: FeaturesFormikValues = {
    featureId: '',
    featureName: '',
    featureEnabled: true,
    featureType: FEATURE_TYPE.SIMPLE,
    aggregationQuery: JSON.stringify(
      { aggregation_name: { sum: { field: 'field_name' } } },
      null,
      4
    ),
    aggregationBy: '',
    aggregationOf: [],
  };

  export const AGGREGATION_TYPES = [
    { value: 'avg', text: 'average()' },
    { value: 'value_count', text: 'count()' },
    { value: 'sum', text: 'sum()' },
    { value: 'min', text: 'min()' },
    { value: 'max', text: 'max()' },
  ];

