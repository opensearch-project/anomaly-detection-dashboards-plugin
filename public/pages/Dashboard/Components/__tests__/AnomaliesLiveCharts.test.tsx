/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { AnomaliesLiveChart } from '../AnomaliesLiveChart';
import { SELECTED_DETECTORS } from '../../../../pages/utils/__tests__/constants';
import { Provider } from 'react-redux';
import { coreServicesMock } from '../../../../../test/mocks';
import { CoreServicesContext } from '../../../../components/CoreServices/CoreServices';
import { mockedStore } from '../../../../redux/utils/testUtils';
const anomalyResponse = [
  {
    ok: true,
    response: {
      anomaly_grade: 0.10949221682655441,
      data_start_time: 1651817250642,
      data_end_time: 1651817310642,
      detector_id: 'gtU2l4ABuV34PY9ITTdm',
      name: 'test2',
    },
  },
];
const anomalyResultQuery = [
  {
    anomaly_grade: 0.10949221682655441,
    data_start_time: 1651817250642,
    data_end_time: 1651817310642,
    detector_id: 'gtU2l4ABuV34PY9ITTdm',
  },
];

jest.mock('../../utils/utils', () => ({
  getLatestAnomalyResultsForDetectorsByTimeRange: jest.fn(
    () => anomalyResponse
  ),
  getFloorPlotTime: jest.fn(() => 1651817250642),
  getLatestAnomalyResultsByTimeRange: jest.fn(() => anomalyResultQuery),
  visualizeAnomalyResultForXYChart: jest.fn(),
}));
describe('<AnomaliesLiveChart /> spec', () => {
  test('AnomaliesLiveChart with Sample anomaly data', async () => {
    const { container, getByTestId, getAllByText, getByText } = render(
      <Provider store={mockedStore()}>
        <CoreServicesContext.Provider value={coreServicesMock}>
          <AnomaliesLiveChart {...SELECTED_DETECTORS} />
        </CoreServicesContext.Provider>
      </Provider>
    );
    //mock current last update to a specific date so doesn't produce new snapshot each minute
    Date.now = jest.fn().mockReturnValue(new Date('2021-06-06T12:33:37.000Z'));
    await waitFor(() => {
      expect(
        getByTestId('dashboardFullScreenButton').innerHTML.includes(
          'euiIcon-isssLoaded'
        )
      );
    });
    await waitFor(() => {});
    expect(container).toMatchSnapshot();
    getAllByText('Detector with the most recent anomaly');
  });
});
