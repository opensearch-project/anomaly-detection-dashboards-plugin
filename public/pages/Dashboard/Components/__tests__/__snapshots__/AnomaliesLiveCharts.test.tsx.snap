// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnomaliesLiveChart /> spec AnomaliesLiveChart with Sample anomaly data 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
          data-test-subj="contentPanelTitle"
        >
          <div
            class="euiFlexItem"
          >
            <h3
              class="euiTitle euiTitle--small"
              data-test-subj="dashboardLiveAnomaliesHeader"
            >
              Live anomalies
               
              <span
                class="euiBadge euiBadge--iconLeft"
                style="background-color: rgb(219, 19, 116); color: rgb(255, 255, 255);"
                title="Live"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    Live
                  </span>
                </span>
              </span>
            </h3>
          </div>
        </div>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem content-panel-subTitle"
            style="line-height: normal; max-width: 75%;"
          >
            Live anomaly results across detectors for the last 30 minutes.  
                'The results refresh every 1 minute. 
                'For each detector, if an anomaly occurrence is detected at the end of the detector interval, 
                'you will see a bar representing its anomaly grade.
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <button
              aria-label="View full screen"
              class="euiButton euiButton--primary"
              data-test-subj="dashboardFullScreenButton"
              type="button"
            >
              <span
                class="euiButtonContent euiButton__content"
              >
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoaded euiButtonContent__icon"
                  focusable="false"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M13 3v4h-1V4H9V3h4ZM3 3h4v1H4v3H3V3Zm10 10H9v-1h3V9h1v4ZM3 13V9h1v3h3v1H3ZM0 1.994C0 .893.895 0 1.994 0h12.012C15.107 0 16 .895 16 1.994v12.012A1.995 1.995 0 0 1 14.006 16H1.994A1.995 1.995 0 0 1 0 14.006V1.994Zm1 0v12.012c0 .548.446.994.994.994h12.012a.995.995 0 0 0 .994-.994V1.994A.995.995 0 0 0 14.006 1H1.994A.995.995 0 0 0 1 1.994Z"
                    fill-rule="evenodd"
                  />
                </svg>
                <span
                  class="euiButton__text"
                >
                  View full screen
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding-top: 175px;"
        >
          <div
            class="euiFlexItem"
            style="min-width: 200px;"
          >
            <div
              class="euiStat euiStat--leftAligned"
            >
              <div
                class="euiText euiText--small euiStat__description"
              >
                <p
                  aria-hidden="true"
                >
                  Last updated time
                </p>
              </div>
              <p
                aria-hidden="true"
                class="euiTitle euiTitle--small euiStat__title"
              >
                06/06/2021 12:33 PM
              </p>
              <p
                class="euiScreenReaderOnly"
              >
                Last updated time 06/06/2021 12:33 PM
              </p>
            </div>
          </div>
          <div
            class="euiFlexItem"
            style="min-width: 310px;"
          >
            <div
              class="euiStat euiStat--leftAligned"
            >
              <div
                class="euiText euiText--small euiStat__description"
              >
                <p
                  aria-hidden="true"
                >
                  Detector with the most recent anomaly
                </p>
              </div>
              <p
                aria-hidden="true"
                class="euiTitle euiTitle--small euiStat__title"
              />
              <p
                class="euiScreenReaderOnly"
              >
                Detector with the most recent anomaly 
              </p>
            </div>
          </div>
          <div
            class="euiFlexItem"
            style="min-width: 185px;"
          >
            <div
              class="euiStat euiStat--leftAligned"
            >
              <div
                class="euiText euiText--small euiStat__description"
              >
                <p
                  aria-hidden="true"
                >
                  Most recent anomaly grade
                </p>
              </div>
              <p
                aria-hidden="true"
                class="euiTitle euiTitle--small euiStat__title"
              >
                0.00e+0
              </p>
              <p
                class="euiScreenReaderOnly"
              >
                Most recent anomaly grade 0.00e+0
              </p>
            </div>
          </div>
        </div>
        <div>
          <div
            style="height: 200px; width: 100%; opacity: 1;"
          >
            <div
              class="echChart"
            >
              <div
                class="echChartBackground"
                style="background-color: transparent;"
              />
              <div
                class="echChartStatus"
                data-ech-render-complete="false"
                data-ech-render-count="0"
              />
              <div
                class="echChartResizer"
              />
              <div
                class="echContainer"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
